src/psyclone/extractor.py

```diff
@@ -0,0 +1,220 @@
+            :type parent: a :py::class::`psyclone.psyGen.Node`.
+        '''
+        Node.__init__(self, children=children, parent=parent)
+
+    def __str__(self):
+        ''' Returns a string representation of the subtree starting at
+        the Extract Node. '''
+        result = "ExtractStart\n"
+        for child in self.children:
+            result += str(child)+"\n"
+        return result+"ExtractEnd"
+
+    @property
+    def coloured_text(self):
+        ''' Returns a string containing the name of this Node along with \
+        control characters for colouring in terminals that supports it.
+
+        :returns: the name of this Node, possibly with control codes for \
+                  colouring.
+        :rtype: string
+        '''
+        return colored("Extract", SCHEDULE_COLOUR_MAP["Extract"])
+
+    @property
+    def dag_name(self):
+        ''' Return the base dag name for this ExtractNode '''
+        return "extract_" + str(self.position)
+
+    def view(self, indent=0):
+        ''' Print a text representation of this Extract schedule to stdout \
+        and then call the view() method of any children. The text \
+        representation returns position of Extract Node(s) in the tree.
+
+        :param int indent: depth of indent for output text.
+        '''
+        print(self.indent(indent) + self.coloured_text +
+              "[position='" + str(self.position) +
+              "',depth='" + str(self.depth) + "']")
+        for entity in self._children:
+            entity.view(indent=indent + 1)
+
+    def gen_code(self, parent):
+        ''' Marks region for code extraction as children of the \
+        ExtractNode. For now it inserts comments at the position of the \
+        ExtractNode and after all children of the ExtractNode. These \
+        comments will later be replaced by calls to write out arguments \
+        of extracted Node(s) or Kernel(s).
+
+        :param parent: the parent of this Node.
+        :type parent: :py:class:`psyclone.psyGen.Node`.
+        '''
+        from psyclone.f2pygen import CommentGen
+        parent.add(CommentGen(parent, ""))
+        parent.add(CommentGen(parent, " ExtractStart"))
+        parent.add(CommentGen(
+            parent, " CALL write_extract_arguments(argument_list)"))
+        parent.add(CommentGen(parent, ""))
+        for child in self.children:
+            child.gen_code(parent)
+        parent.add(CommentGen(parent, ""))
+        parent.add(CommentGen(parent, " ExtractEnd"))
+        parent.add(CommentGen(parent, ""))
+
+
+class Extractor(object):
+    ''' This class contains helper functions for code extraction. For \
+    now it only provides the function to extract the specific Kernel \
+    from an Invoke Schedule. Another planned functionality is to wrap \
+    settings for generating driver for the extracted code. '''
+
+    @staticmethod
+    def extract_kernel(schedule, kernel_name, root_node_position=None):
```

Comment:
I'm sorry but (unless I've missed something) I'm really not convinced that this routine is needed. I think PSyKE is a tool for the HPC expert who will need to extract a kernel in the context of the PSy layer as a whole - i.e. including any OpenMP or OpenACC directives. The safest and most robust way to do this is to look at the Schedule and select the set of Nodes to extract. I think starting from the Kernel and going up is making life difficult as you need to allow for all sorts of things that could be an ancestor of the kernel (e.g. an OpenMP parallel directive or an OpenACC data region). This is highlighted by the fact that you've had to add a `depth` property to the Node - something that it should not have to care about since it cannot reliably be used to reason about things (since that requires us to predict all possible positions of a Node in a Schedule).

Reply:
Removed the routine. As the `Extractor` class became empty I removed it as well. I will add it back in issue #298 as I will need it to wrap-up driver settings. For now I added the explanation in the module description with the issue number.